crates/solidity/inputs/language/bindings/rules.msgb

```diff
@@ -6,178 +8,326 @@ attribute push_symbol = symbol       => type = "push_symbol", symbol = symbol
-;;; Interface definitions
+;;; Named definitions (contracts, functions, libraries, etc.)
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-@interface [InterfaceDefinition] {
-  node @interface.lexical_scope
-  node @interface.defs
+@contract [ContractDefinition ... @name name: [Identifier] ...] {
+  node def
+  attr (def) node_definition = @name
+
+  edge @contract.def -> def
+
+  node type_def
+  attr (type_def) pop_symbol = "@typeof"
+
+  node member
+  attr (member) pop_symbol = "."
+
+  edge def -> type_def
+  edge type_def -> member
+  edge member -> @contract.members
+
+  node type_member
+  attr (type_member) pop_symbol = "."
+  edge def -> type_member
+
+  edge type_member -> @contract.type_members
 }
 
 @interface [InterfaceDefinition ... @name name: [Identifier] ...] {
   node def
   attr (def) node_definition = @name
 
-  edge @interface.lexical_scope -> def
-  edge @interface.defs -> def
-}
+  edge @interface.def -> def
 
-;; Connect the interface to its containing source unit
-@source_unit [SourceUnit [SourceUnitMembers
-    ...
-    [SourceUnitMember @interface [InterfaceDefinition]]
-    ...
-]] {
-  edge @source_unit.defs -> @interface.defs
-  edge @interface.lexical_scope -> @source_unit.lexical_scope
-}
+  node type_def
+  attr (type_def) pop_symbol = "@typeof"
 
+  node member
+  attr (member) pop_symbol = "."
 
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;;; Library definitions
-;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  edge def -> type_def
+  edge type_def -> member
+  edge member -> @interface.members
 
-@library [LibraryDefinition] {
-  node @library.lexical_scope
-  node @library.defs
+  node type_member
+  attr (type_member) pop_symbol = "."
+  edge def -> type_member
+
+  edge type_member -> @interface.type_members
 }
 
 @library [LibraryDefinition ... @name name: [Identifier] ...] {
   node def
   attr (def) node_definition = @name
 
-  edge @library.lexical_scope -> def
-  edge @library.defs -> def
+  edge @library.def -> def
+
+  node member
+  attr (member) pop_symbol = "."
+  edge def -> member
+
+  edge member -> @library.members
 }
 
-;; Connect the library to its containing source unit
-@source_unit [SourceUnit [SourceUnitMembers
-    ...
-    [SourceUnitMember @library [LibraryDefinition]]
-    ...
-]] {
-  edge @source_unit.defs -> @library.defs
-  edge @library.lexical_scope -> @source_unit.lexical_scope
+@function [FunctionDefinition ... name: [FunctionName ... @name [Identifier] ...] ...] {
+  node def
+  attr (def) node_definition = @name
+
+  edge @function.def -> def
 }
 
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
-;;; Function definitions
+;;; Type names
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-@function [FunctionDefinition] {
-  node @function.lexical_scope
-  node @function.defs
+@type_name [TypeName] {
```

Comment:
Question:
IIUC, the reason we split every definition here into two blocks, is that the evaluation of this `msgb` file is "strict", vs "lazy", which allows us to refer to definitions later in the file. And the main reason we are using "strict" evaluation is speed. Is that correct? or is it also correctness?

I wonder what is the speed impact here, given that we only evaluate the file only once, and cache it in memory.

Just trying to understand here. Not proposing we make any changes.

Reply:
Hmm... actually we're [using lazy evaluation](https://github.com/manastech/slang/blob/feat%2Fbindings-assertions/crates/metaslang/bindings/src/builder/mod.rs#L398) when building the stack graph (this code was taken from `tree-sitter-stack-graphs`). We _were_ using strict evaluation when building the graph manually for testing and snapshots, but that code no longer exists.

Anyway, yes, the reason for some of the splits are to allow strict execution, where the stanzas are executed in order and you cannot reference a node without first declaring it.

I thought strict evaluation should be faster and more efficient overall, but the [original tree-sitter documentation](https://docs.rs/tree-sitter-graph/latest/tree_sitter_graph/reference/index.html#high-level-structure) says the opposite.

> Regular execution will apply the stanzas in order, and it is important to make sure that scoped variables have been assigned before they are used. This is not a requirement when using the lazy evaluation strategy, which handles this implicitly. The lazy evaluation strategy is also more efficient when there are many stanzas, because it can reduce tree traversals. Therefore, using the lazy evaluation strategy is recommended, and will likely become the only supported strategy in future releases.

I'm not sure _how_ using lazy evaluation could lead to less tree traversals. Compare [strict](https://github.com/manastech/slang/blob/feat/bindings-assertions/crates/metaslang/graph_builder/src/execution/strict.rs#L56-L69) vs [lazy](https://github.com/manastech/slang/blob/feat/bindings-assertions/crates/metaslang/graph_builder/src/execution/lazy.rs#L59-L91).

In any case, it shouldn't matter for correctness.

Reply:
I see. In that case (for future revisions), do you think it is more readable/maintainable to define all edges/nodes for a specific language element in one block? or is there another benefit to keep them separate?

Reply:
I'm not sure what should be the general rule here. In most cases, I think it's clearer if the nodes that connect to the rest of the graph are declared in a separate stanza, especially when there are multiple stanzas covering different use cases. For example:
```
@param [Parameter] {
  node @param.lexical_scope
  node @param.def
}

@param [Parameter ... @type_name [TypeName] ...] {
  edge @type_name.type_ref -> @param.lexical_scope
}

@param [Parameter ... @type_name [TypeName] ... @name [Identifier]] {
  node def
  attr (def) node_definition = @name

  edge @param.def -> def

  node typeof
  attr (typeof) push_symbol = "@typeof"

  edge def -> typeof
  edge typeof -> @type_name.output
}
```
Here we will use `@param.lexical_scope` and `@param.def` to connect to the rest of the graph. And even though the second stanza applies to all parameters (ie. even for named parameters), so those nodes could be defined there, I think it's clearer to have them separate.

In other cases, like for the top-level entities, it makes sense to define them separately to be able to have proper variable names associated to the graph nodes, for debugging and inspecting a graph.

There have been some instances where I got a bit confused because I couldn't immediately find the nodes defined from a CST node, but I'd say those have been the minority.

Reply:
Makes sense. Thanks for explaining.